edb/lang/edgeql/compiler/pathctx.py

```diff
@@ -91,14 +91,9 @@ def set_path_alias(
 
 
 def infer_cardinality(
-        expr: irast.Base, *, ctx: context.ContextLevel) -> irast.Cardinality:
-    return irinference.infer_cardinality(expr, ctx.path_scope, ctx.schema)
-
-
-def enforce_singleton(expr: irast.Base, *, ctx: context.ContextLevel) -> None:
-    cardinality = infer_cardinality(expr, ctx=ctx)
-    if cardinality != irast.Cardinality.ONE:
-        raise errors.EdgeQLError(
-            'possibly more than one element returned by an expression '
-            'where only singletons are allowed',
-            context=expr.context)
+        expr: irast.Base, *,
+        scope: typing.Optional[irast.ScopeTreeNode]=None,
+        ctx: context.ContextLevel) -> irast.Cardinality:
+    if scope is None:
+        scope = ctx.path_scope
```

Comment:
I'd have "scope" and "schema" parameters (so no "ctx") for this function.

Reply:
At this point it devolved into a pointless wrapper around the general `infer_cardinality`, so I'll just drop it altogether.